lib/osutil/osutil.go

```diff
@@ -26,12 +26,54 @@ var renameLock = sync.NewMutex()
 // TryRename renames a file, leaving source file intact in case of failure.
 // Tries hard to succeed on various systems by temporarily tweaking directory
 // permissions and removing the destination file when necessary.
-func TryRename(filesystem fs.Filesystem, from, to string) error {
+func TryRename(from fs.Filesystem, fromPath string, to fs.Filesystem, toPath string) error {
 	renameLock.Lock()
 	defer renameLock.Unlock()
 
-	return withPreparedTarget(filesystem, from, to, func() error {
-		return filesystem.Rename(from, to)
+	return withPreparedTarget(from, fromPath, to, toPath, func() error {
+		// Optimization
+		if from == to || to == nil || (from.Type() == to.Type() && from.URI() == to.URI()) {
+			return from.Rename(fromPath, toPath)
+		}
+
+		// See if one is a prefix of the other filesystem.
+		if from.Type() == to.Type() {
+			fromUri := from.URI()
+			toUri := to.URI()
+			if strings.HasPrefix(toUri, fromUri) {
+				newToPath := filepath.Join(strings.TrimPrefix(toUri, fromUri), toPath)
+				if err := from.Rename(fromPath, newToPath); err == nil {
+					return err
+				}
+			} else if strings.HasPrefix(fromPath, toPath) {
+				newFromPath := filepath.Join(strings.TrimPrefix(fromUri, toUri), fromPath)
+				if err := to.Rename(newFromPath, toPath); err == nil {
+					return err
+				}
+			} else {
```

Comment:
This whole thing in general doesn't give me a lot of confidence. 

Was thinking about cases where it's C drive vs D drive, or /a/b/c renamed to /a/b/../../d. In blank